### PR TITLE
[Spark][Version Checksum] Read Protocol, Metadata, and ICT directly from the Checksum during Snapshot construction

### PR DESCRIPTION
**Pull request by dhruvarya-db**
#### Which Delta project/connector is this regarding?

- [X] Spark
- [ ] Standalone
- [ ] Flink
- [ ] Kernel
- [ ] Other (fill in here)

## Description

Stacked over https://github.com/delta-io/delta/pull/3907. 
This PR makes the Checksum (if available) the source of truth for Protocol, Metadata, ICT during snapshot construction. This helps us avoid a Spark query and improves performance.

## How was this patch tested?

Added some test cases to existing suites

## Does this PR introduce _any_ user-facing changes?

No

### CODE REVIEW

File: spark/src/main/scala/org/apache/spark/sql/delta/Checksum.scala [hunk 1]
```diff
@@ -795,8 +795,11 @@ trait ValidateChecksum extends DeltaLogging { self: Snapshot =>
       }
     }
 
-    compareAction(checksum.metadata, computedStateToCheckAgainst.metadata, "Metadata", "metadata")
-    compareAction(checksum.protocol, computedStateToCheckAgainst.protocol, "Protocol", "protocol")
+    if (spark.sessionState.conf.getConf(
+        DeltaSQLConf.USE_PROTOCOL_AND_METADATA_FROM_CHECKSUM_ENABLED)) {
```
Comment (scovich) [inline spark/src/main/scala/org/apache/spark/sql/delta/Checksum.scala:799]: Is it useful/important to make this check conditional, where it was previously unconditional?
What is the rationale?

File: spark/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala [hunk 2]
```diff
@@ -377,6 +377,41 @@ class Snapshot(
...
+          val metadataStr = Option(m).map(_.toString).getOrElse("null")
+          recordDeltaEvent(
+            deltaLog,
+            opType = "delta.assertions.missingEitherProtocolOrMetadataFromChecksum",
+            data = Map(
+              "version" -> version.toString, "protocol" -> protocolStr, "source" -> metadataStr))
+          logWarning(log"Either protocol or metadata is null from checksum; " +
+            log"version:${MDC(DeltaLogKeys.VERSION, version)} " +
+            log"protocol:${MDC(DeltaLogKeys.PROTOCOL, protocolStr)} " +
+            log"metadata:${MDC(DeltaLogKeys.DELTA_METADATA, metadataStr)}")
```
Comment (scovich) [inline spark/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala:408]: Why is the one called `PROTOCOL` but the other `DELTA_METADATA`? Why not just `METADATA`?

Comment (scovich) [reply spark/src/main/scala/org/apache/spark/sql/delta/Snapshot.scala:408]: update: There are already several `METADATA_xxx` keys, I guess this avoids ambiguity?

File: spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala [hunk 3]
```diff
@@ -179,6 +179,8 @@ class InCommitTimestampSuite
   }
 
   test("Missing CommitInfo should result in a DELTA_MISSING_COMMIT_INFO exception") {
+    // Make sure that we don't retrieve the time from the CRC.
+    withSQLConf(DeltaSQLConf.DELTA_WRITE_CHECKSUM_ENABLED.key -> "false") {
     withTempDir { tempDir =>
```
Comment (scovich) [inline spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala:184]: odd (lack of) indentation here?

(several more below)

**Review by scovich — APPROVED** [review]
Couple nits, but code itself LGTM

File: spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala (hunk 3 above)
Comment (dhruvarya-db) [reply spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala]: Fixed the indentation. I wanted to keep the diff cleaner but given that the content itself has been reviewed, that should not matter anymore

File: spark/src/main/scala/org/apache/spark/sql/delta/Checksum.scala (hunk 1 above)
Comment (dhruvarya-db) [reply spark/src/main/scala/org/apache/spark/sql/delta/Checksum.scala]: I though about this again, and this does not make much sense. Removed

File: spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala (hunk 3 above)
Comment (scovich) [reply spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala]: FYI - Indentation changes are easily hidden:
image

Comment (scovich) [reply spark/src/test/scala/org/apache/spark/sql/delta/InCommitTimestampSuite.scala]: ... which gives e.g.
image

**Pull request merged**